src/compiler/checker.ts

```diff
@@ -10985,7 +10985,61 @@ namespace ts {
             return -1;
         }
 
-        function isInLegalTypePredicatePosition(node: Node): boolean {
+        function checkTypePredicate(node: TypePredicateNode) {
+            const parent = getTypePredicateParent(node);
+            if (!parent) {
+                return;
+            }
+            const typePredicate = getSignatureFromDeclaration(parent).typePredicate;
+            if (typePredicate.parameterIndex >= 0) {
+                if (parent.parameters[typePredicate.parameterIndex].dotDotDotToken) {
+                    error(node.parameterName,
+                        Diagnostics.A_type_predicate_cannot_reference_a_rest_parameter);
+                }
+                else {
+                    checkTypeAssignableTo(typePredicate.type,
+                        getTypeOfNode(parent.parameters[typePredicate.parameterIndex]),
+                        node.type);
+                }
+            }
+            else if (node.parameterName) {
+                let hasReportedError = false;
+                for (var param of parent.parameters) {
+                    if (hasReportedError) {
+                        break;
+                    }
+                    if (param.name.kind === SyntaxKind.ObjectBindingPattern ||
+                        param.name.kind === SyntaxKind.ArrayBindingPattern) {
+
+                        (function checkBindingPattern(pattern: BindingPattern) {

```

Comment:
You can extract this out of the function and call it to reduce nesting and avoid a closure allocation per invocation.